x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/PutSamlServiceProviderRequest.java

```diff
@@ -0,0 +1,127 @@
+        final SamlServiceProviderDocument document = SamlServiceProviderDocument.fromXContent(null, parser);
+        if (document.entityId == null) {
+            document.setEntityId(entityId);
+        } else if (entityId != null) {
+            if (entityId.equals(document.entityId) == false) {
+                throw new ElasticsearchParseException(
+                    "Entity id [{}] inside request body and entity id [{}] from parameter do not match", document.entityId, entityId);
+            }
+        }
+        if (document.created != null) {
+            throw new ElasticsearchParseException(
+                "Field [{}] may not be specified in a request", SamlServiceProviderDocument.Fields.CREATED_DATE);
+        }
+        if (document.lastModified != null) {
+            throw new ElasticsearchParseException(
+                "Field [{}] may not be specified in a request", SamlServiceProviderDocument.Fields.LAST_MODIFIED);
+        }
+        document.setCreatedMillis(System.currentTimeMillis());
+        document.setLastModifiedMillis(System.currentTimeMillis());
+        return new PutSamlServiceProviderRequest(document);
+    }
+
+    public PutSamlServiceProviderRequest(SamlServiceProviderDocument document) {
+        this.document = document;
+    }
+
+    public PutSamlServiceProviderRequest(StreamInput in) throws IOException {
+        this.document = new SamlServiceProviderDocument(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        document.writeTo(out);
+    }
+
+    public SamlServiceProviderDocument getDocument() {
+        return document;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        final ValidationException docException = document.validate();
+        ActionRequestValidationException validationException = null;
+        if (docException != null) {
+            validationException = new ActionRequestValidationException();
+            validationException.addValidationErrors(docException.validationErrors());
+        }
+
+        if (Strings.hasText(document.acs)) { // if this is blank the document validation will fail
+            try {
+                final URL url = new URL(document.acs);
+                if (url.getProtocol().equals("https") == false) {
+                    validationException = addValidationError(
+                        "[" + SamlServiceProviderDocument.Fields.ACS + "] must be a [https] URL", validationException);
+                }
+            } catch (MalformedURLException e) {
+                String error = "[" + SamlServiceProviderDocument.Fields.ACS + "] must be a valid URL";
+                if (e.getMessage() != null) {
+                    error += " - " + e.getMessage();
+                }
+                validationException = addValidationError(error, validationException);
+            }
+        }
+
+        if (document.certificates.identityProviderSigning.isEmpty() == false) {
```

Comment:
Because we have no other way to validate that correct data is sent here.

I could accept it but enforce that it can only be the same certificate that is configured for the IDP, but that seems worse to me. I'm happy to discuss that option if you prefer it, but to my thinking "you can provide this field, but only with this value that is pre-configured in the YML and you don't have access to" is a pretty weird API.
And it forces us to continue to support that field in the API even if we decide that we want a different approach (like implicit generation of signing certs) down the track.
 
I can't accept any other certificates, because we wont have the keys for them, so it would be allowing the creation of a broken SP.

Actually my preference would be for this field to go away entirely, because we don't know how to support it yet, and I feel like I made the wrong choice by supporting it in the index document before we really know how it will be handled. But I took the less invasive approach and just decided that we don't support it for SP registration yet.

Reply:
What I was thinking about wouldn't need to do validation, but it would need the private keys to be passed in the API request and stored _somewhere_ and we don't have this yet. 
This is the second time I make a similar comment without taking the private keys into consideration :man_facepalming: . 

Given that we will not store these in the same index ever ,I'm ok with leaving this as is here - if we were writing this now from scratch I'd remove this altogether as you say and handle it only as yml settings until we have a better story for per SP signing credentials